lte/gateway/c/core/oai/test/amf/test_amf_procedures.cpp

```diff
@@ -2472,4 +2554,85 @@ TEST_F(AMFAppProcedureTest, TestAuthFailureFromSubscribeDbLock) {
   amf_free_ue_context(ue_context_p);
 }
 
+TEST_F(AMFAppProcedureTest, TestPDUSession_LocationUpdateFail) {
+  int rc = RETURNerror;
+  amf_ue_ngap_id_t ue_id = 0;
+  std::vector<MessagesIds> expected_Ids{
+      AMF_APP_NGAP_AMF_UE_ID_NOTIFICATION,  // new registration notification
+                                            // indication to ngap
+      NGAP_NAS_DL_DATA_REQ,                 // Authentication Request to UE
+      NGAP_NAS_DL_DATA_REQ,  // Security Command Mode Request to UE
+      NGAP_NAS_DL_DATA_REQ,  // PDU Session Establishment Request with failure
+                             // sm cause
+      NGAP_NAS_DL_DATA_REQ,  // Deregistaration Accept
+      NGAP_UE_CONTEXT_RELEASE_COMMAND  // UEContextReleaseCommand
+  };
+
+  /* Send the initial UE message */
+  imsi64_t imsi64 = 0;
+  imsi64 = send_initial_ue_message_no_tmsi(amf_app_desc_p, 36, 1, 1, 0, plmn,
+                                           initial_ue_message_hexbuf,
+                                           sizeof(initial_ue_message_hexbuf));
+
+  /* Check if UE Context is created with correct imsi */
+  EXPECT_TRUE(get_ue_id_from_imsi(amf_app_desc_p, imsi64, &ue_id));
+
+  /* Send the authentication response message from subscriberdb */
+  rc = send_proc_authentication_info_answer(imsi, ue_id, true);
+  EXPECT_TRUE(rc == RETURNok);
+
+  /* Send uplink nas message for auth response from UE */
+  rc = send_uplink_nas_message_ue_auth_response(
+      amf_app_desc_p, ue_id, plmn, ue_auth_response_hexbuf,
+      sizeof(ue_auth_response_hexbuf));
+  EXPECT_TRUE(rc == RETURNok);
+
+  /* Send uplink nas message for security mode complete response from UE */
+  rc = send_uplink_nas_message_ue_smc_response(amf_app_desc_p, ue_id, plmn,
+                                               ue_smc_response_hexbuf,
+                                               sizeof(ue_smc_response_hexbuf));
+  EXPECT_TRUE(rc == RETURNok);
+
+  s6a_update_location_ans_t ula_ans = util_amf_send_s6a_ula(imsi);
+  strcpy(ula_ans.imsi, "123456789012345");
+  rc = amf_handle_s6a_update_location_ans(&ula_ans);
+  EXPECT_EQ(rc, RETURNerror);
+
+  send_initial_context_response(amf_app_desc_p, ue_id);
+
+  /* Send uplink nas message for registration complete response from UE */
+  rc = send_uplink_nas_registration_complete(
+      amf_app_desc_p, ue_id, plmn, ue_registration_complete_hexbuf,
+      sizeof(ue_registration_complete_hexbuf));
+  EXPECT_TRUE(rc == RETURNok);
+
+  ue_m5gmm_context_t* ue_context_p =
+      amf_ue_context_exists_amf_ue_ngap_id(ue_id);
+  // ue context should exist
+  ASSERT_NE(ue_context_p, nullptr);
+  memset(&ue_context_p->amf_context.apn_config_profile, 0,
+         sizeof(ue_context_p->amf_context.apn_config_profile));
+
+  /* Send uplink nas message for pdu session establishment request from UE */
+  rc = send_uplink_nas_pdu_session_establishment_request(
```

Comment:
Can set expect for PDU session reject message

Reply:
// smf context should not be present
EXPECT_EQ(ue_context_p->amf_context.smf_ctxt_map.size(), 0);

We are checking here for smf context has been created or not, if pdu session establishment request is rejected then the smf_ctxt_map.size() should be zero.

Reply:
@rsarwad, we are defining event IDs to check expected events are happening or not.

std::vector<MessagesIds> expected_Ids{
      AMF_APP_NGAP_AMF_UE_ID_NOTIFICATION,  // new registration notification indication to ngap
      NGAP_NAS_DL_DATA_REQ,                 // Authentication Request to UE
      NGAP_NAS_DL_DATA_REQ,  // Security Command Mode Request to UE
      NGAP_NAS_DL_DATA_REQ,  // PDU Session Establishment Request with failure sm cause
      NGAP_NAS_DL_DATA_REQ,  // Deregistaration Accept
      NGAP_UE_CONTEXT_RELEASE_COMMAND  // UEContextReleaseCommand
  };
  
  we are checking the expected events and actual events are matching or not here
  EXPECT_EQ(expected_Ids, AMFClientServicer::getInstance().msgtype_stack);
  
After PDU session establishment request, we are checking smf context is present or not.
Case 1 : If context is present then pdu establishment is accepted and AMF sends downlink message to NGAP.
Case 2 : If context is not present then pdu establishment is rejected and AMF sends downlink message to NGAP.
Case 3 : If message dropped, there is no context and AMF does not send downlink message to AMF.

In this test, there is no smf context and AMF sending downlink to NGAP, means the downlink message is PDU session establishment reject.